## Data Cleaning

In this Notebook, we will be using a revised version of the [Pokemon Dataset](https://www.kaggle.com/mariotormo/complete-pokemon-dataset-updated-090420), purposefully dirtied for the purpose of performing data cleaning. Additionally, some columns were also omitted for this notebook.

Our Notebooks are designed to be guided learning activities. To use them, simply go through the cells from top to bottom, following the directions along the way. If you find any unclear parts or mistakes in the Notebooks, email your instructor.

## Instructions
* Read each cell and implement the TODOs sequentially. The markdown/text cells also contain instructions which you need to follow to get the whole notebook working.
* Do not change the variable names unless the instructor allows you to.
* You are expected to search how to some functions work on the Internet or via the docs.
* The notebooks will undergo a 'Restart and Run All' command, so make sure that your code is working properly.
* You are expected to understand the dataset loading and processing separately from this class.
* You may not reproduce this notebook or share them to anyone.

## Import
Import **numpy** and **pandas**.

[**`pandas`**](https://pandas.pydata.org/pandas-docs/stable/index.html) is a software library for Python which provides data structures and data analysis tools.

In [1]:
import numpy as np
import pandas as pd

## The Dataset

The Pokemon dataset contains 890 known pokemon until 8th Generation and its varieties.

The dataset is provided to you as a `.csv` file. `.csv` means comma-separated values. You can open the file in Notepad to see how it is exactly formatted.

If you view the `.csv` file in Excel, you can see that our dataset contains many **observations** (rows) across 20 **variables** (columns). An observation corresponds to a Pokemon (or a variant of it), and the variables correspond to the attributes of that Pokemon. We will describe what some of the columns mean as we go through this notebook.

## Reading the Dataset
Our first step is to load the dataset using pandas. This will load the dataset into a pandas `DataFrame`. To load the dataset, we use the [`read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function. Note that you may need to change the path depending on the location of the file in your machine.

In [2]:
pokemon_df = pd.read_csv('pokedex.csv')

Whenever we load a new dataset, it is generally a good idea to call the [`info()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) function, which displays general information about the dataset.

In [3]:
pokemon_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1028 entries, 0 to 1027
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        1028 non-null   int64  
 1   pokedex_number    1028 non-null   int64  
 2   name              1028 non-null   object 
 3   generation        1028 non-null   int64  
 4   status            1028 non-null   object 
 5   species           1028 non-null   object 
 6   type_number       1028 non-null   int64  
 7   type_1            1028 non-null   object 
 8   type_2            542 non-null    object 
 9   height_m          1028 non-null   float64
 10  weight_kg         1027 non-null   float64
 11  abilities_number  1028 non-null   int64  
 12  ability_1         1025 non-null   object 
 13  total_points      1028 non-null   int64  
 14  hp                921 non-null    float64
 15  attack            1026 non-null   float64
 16  defense           842 non-null    float64


## Cleaning the Dataset
Before we can start exploring our dataset, we have to clean it first in case there are inconsistencies that may incur problems in analysis.

Throughout the notebook, we will look at certain scenarios and see whether we should clean the relevant variables or not.

## Scenario:

Let's say that we want to see **the distribution of the pokemons' statuses for a given generation**. For this, we would use the variables:
- **`generation`**: numbered generation which the Pokemon was first introduced
- **`status`**: denotes if the Pokemon is normal, sub legendary, legendary or mythical. (We can think of this as the Pokemon's group)

Thus, we should inspect these columns and see if these need cleaning.

### `generation` variable
The pokemon series has 8 generations as of date, with the newest generation being Pokemon Sword and Shield. As such, we can expect that the possible values for the `generation` column are the whole numbers 1 to 8.
Use the [`unique()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html) function to list all unique values for this variable.

In [4]:
pokemon_df['generation'].unique()

array([1, 2, 3, 4, 5, 6, 7, 8])

We can see here that the unique values for the `generation` variable are the numbers 1 to 8 -- all of which are valid.

### `status` variable
As described earlier, the valid values for the `status` variable are either 'Normal', 'Sub Legendary', 'Legendary', or 'Mythical'. List all the unique values for the `status` variable.

In [5]:
# Write your code here
pokemon_df['status'].unique()

array(['Normal', 'Sub Legendary', 'Legendary', 'Mythical'], dtype=object)

Since all of the values are valid, we do not need to clean the `status` variable.

## Scenario:
Let's say that we want to explore the different [types](https://bulbapedia.bulbagarden.net/wiki/Type) of Pokemon. You can think of 'types' as 'elements' in the Pokemon Universe. Generally, a pokemon can have up to two types: a primary type and an optional secondary type.

For this, we would probably use the following columns:
- **`type_1`**: primary type of the Pokemon
- **`type_2`**: secondary type of the Pokemon (if any)

Thus, we should inspect these and see if these columns need cleaning.

### `type_1` variable
The `type_1` variable is the primary type of the Pokemon. Let's list all the unique values for the `type_1` variable.

In [6]:
pokemon_df['type_1'].unique()

array(['Grass', 'Fire', 'water', 'Bug', 'normal', 'Dark', 'Normal',
       'Poison', 'Electric', 'Grund', 'Ice', 'Fairy', 'Ground', 'Steel',
       'Water', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Dragon', 'watr',
       'Flying'], dtype=object)

Currently, there are 18 unique types in the main Pokemon franchise. You may refer to this [page](https://bulbapedia.bulbagarden.net/wiki/Type) to see the valid list of types in the main pokemon franchise Looking at the results above, we can see that there are multiple representations of the same type (`Water`, `water` for the Water type; `normal`, `Normal` for the Normal type) and there are also some values that have spelling/encoding errors (`watr` - a typo of the Water type; `Grund` - a typo of the Ground type).

Generally speaking, if you are unsure on what the correct representation is, or there isn't really a *correct* representation per se, we can use the most prevalent representation in the variable. To find out the number of observations per unique value, we use the [`value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html) function. This function returns the count of each unique value in the Series.

In [7]:
pokemon_df['type_1'].value_counts()

type_1
Water       112
Grass        91
Normal       89
Bug          81
Psychic      76
Fire         65
Electric     61
Rock         60
Dark         44
Ghost        41
Dragon       40
Poison       39
Fighting     38
Ice          36
Steel        36
Ground       31
normal       26
Fairy        22
water        15
Grund        10
Flying        8
watr          7
Name: count, dtype: int64

From the results above, we can see that `water` and `normal` are less prevalent than `Water` and `Normal`. Thus, it might be more preferrable to use `Water` and `Normal` to represent the Water and Normal types, respectively.

We use the [`replace()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html) function to change the values to the correct representation.

In [8]:
pokemon_df['type_1'] = pokemon_df['type_1'].replace({'water': 'Water', 'normal': 'Normal',
                                                     'watr':'Water', 'Grund': 'Ground'})

Once again, to check if the values were mapped correctly, we can confirm the counts through the `value_counts()` function.

In [9]:
pokemon_df['type_1'].value_counts()

type_1
Water       134
Normal      115
Grass        91
Bug          81
Psychic      76
Fire         65
Electric     61
Rock         60
Dark         44
Ground       41
Ghost        41
Dragon       40
Poison       39
Fighting     38
Ice          36
Steel        36
Fairy        22
Flying        8
Name: count, dtype: int64

Counts of each value of the `type_1` variable after cleaning are shown above.

### `type_2` variable

The `type_2` variable is the secondary type of the Pokemon. Let's list all the unique values for the `type_2` variable. Note that the secondary type is optional, hence it is okay to have empty values (i.e., `nan`).

In [10]:
pokemon_df['type_2'].unique()

array(['Poison', nan, 'Flying', 'Dragon', 'Normal', 'normal', 'Psychic',
       'Steel', 'Ground', 'Fairy', 'Grass', 'Fighting', 'Electric', 'Ice',
       'Dark', 'Ghost', 'Rock', 'Water', 'Poson', 'Fire', 'Bug', 'grass'],
      dtype=object)

Let's check the number of observations per value in the `type_2` variable before cleaning it.

In [11]:
# Write your code here
pokemon_df['type_2'].value_counts()

type_2
Flying      109
Fairy        41
Ground       39
Psychic      38
Steel        35
Fighting     32
Poison       32
Dragon       32
Dark         26
Ghost        25
Grass        23
Water        19
Ice          19
Fire         17
Rock         15
Electric     11
Bug           9
Normal        8
Poson         6
normal        3
grass         3
Name: count, dtype: int64

Identify and address multiple representations and spelling errors then write your code in the cell below.

*You may use either of the approaches presented earlier.*

In [15]:
# Write your code here
pokemon_df['type_2'] = pokemon_df['type_1'].replace({'Poson': 'Poison', 'normal': 'Normal', 'grass':'Grass'})

After addressing multiple representation and spelling errors, let's check again the number of unique observations per value in the `type_2` variable.

In [16]:
# Write your code here
pokemon_df['type_2'].value_counts()

type_2
Water       134
Normal      115
Grass        91
Bug          81
Psychic      76
Fire         65
Electric     61
Rock         60
Dark         44
Ground       41
Ghost        41
Dragon       40
Poison       39
Fighting     38
Ice          36
Steel        36
Fairy        22
Flying        8
Name: count, dtype: int64

Counts of each value of the `type_2` variable after cleaning are shown above.

## Checking for `NaN`s
We will now handle missing data in our dataset. Let us check each variable if it contains a `NaN` / `null` value. We'll use the [`isnull()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html) and [`any()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.any.html) functions for this. This function call will list each variable with a boolean value indicating if the variable contains a `NaN` / `null` value.

In [18]:
pokemon_df.isnull().any()

Unnamed: 0          False
pokedex_number      False
name                False
generation          False
status              False
species             False
type_number         False
type_1              False
type_2              False
height_m            False
weight_kg            True
abilities_number    False
ability_1            True
total_points        False
hp                   True
attack               True
defense              True
sp_attack            True
sp_defense           True
speed                True
dtype: bool

Let's get the list of columns with `NaN` / `null` values.

In [19]:
nan_variables = pokemon_df.columns[pokemon_df.isnull().any()].tolist()
print(nan_variables)

['weight_kg', 'ability_1', 'hp', 'attack', 'defense', 'sp_attack', 'sp_defense', 'speed']


For each variable, show the total number of observation with `NaN` / `null` value. The [`sum()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) function is used to get the counts of null values per column. This works because the [`isnull()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html) function converts the value of the cell to either `True` (if the value is `null` or `nan`) or `False` (if there is a value in the cell). Just like in most programming languages, `True` is equivalent to `1` when used in place of an integer.

In [20]:
pokemon_df[nan_variables].isnull().sum()

weight_kg       1
ability_1       3
hp            107
attack          2
defense       186
sp_attack       3
sp_defense      1
speed           2
dtype: int64

Variables with `NaN` values are shown above.

The [`shape`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html) attribute returns the dimensions of the dataframe.

In [21]:
pokemon_df.shape

(1028, 20)

Before cleaning, we can see here that there are a total of 1028 observations in our dataset.

The `type_2` variable have several observations that are missing a value but it is alright in this context, as having a secondary type is optional for a Pokemon. We cannot simply drop Pokemons with missing values for `hp` and `defense` since there is a significant number of observations with missing values for these variables. Thus, we need to exclude them and find a better way to deal with these missing values.

In [22]:
exclude_variables = ['type_2', 'hp', 'defense']
var_to_drop = list(set(nan_variables) - set(exclude_variables))
print(var_to_drop)

['weight_kg', 'attack', 'sp_defense', 'ability_1', 'sp_attack', 'speed']


Since we see that there is a very small number of observation with `NaN` / `null` value for the `ability_1`, `weight_kg`, `sp_defense`, `sp_attack`, `attack`, and `speed` variables, we can decide to drop those observations from the dataset.

However, remember that in this example, we're dropping these as an example for dropping `null` values. Depending on the situation, you may have to assess whether or not these variables are important.

**Hint!** Use the [`dropna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) function (and its subset parameter).

In [23]:
# Write your code here
pokemon_df = pokemon_df.dropna(subset=var_to_drop)


Display the total number of observations left after removing observations.

In [ ]:
# Write your code here
pokemon_df.shape


weight_kg       0
ability_1       0
hp            106
attack          0
defense       184
sp_attack       0
sp_defense      0
speed           0
dtype: int64

After dropping observations with `NaN` values, we were left with fewer number of observations.

Check if the observations with `NaN` / `null` values for the `weight_kg` variable have been dropped from the dataset. Make sure that the `NaN` value is not included in the list of all values for the `weight_kg` variable.

In [34]:
# Write your code here
bool(pokemon_df['weight_kg'].isnull().any())

False

**Sanity Check!** Your code should return the output:

```False```

We can also check the counts for all of our initial variables with missing values using the code below:

In [57]:
pokemon_df[nan_variables].isnull().sum()

type_2        486
weight_kg       1
ability_1       3
hp            107
attack          2
defense       186
sp_attack       3
sp_defense      1
speed           2
dtype: int64

### `NaN`s in the `type_2` variable

There is a significant number of observations with `NaN` / `null` value for the `type_2` variable. However, we know that missing values for `type_2` are intended (as opposed to randomly missing) since having a secondary type is optional for a Pokemon. As such, instead of removing these observations from the dataset, we can just set a *sentinel value* to identify these observations. For this, we can just set the string `'None'` as our sentinel value. Use the [`loc()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) function to select and assign `'None'` as value for these variables of these observations

In [35]:
pokemon_df.loc[pokemon_df['type_2'].isnull(), 'type_2'] = 'None'

Let's count the number of observations with the sentinel value

In [36]:
pokemon_df.loc[pokemon_df['type_2'] == 'None', 'type_2'].shape

(0,)

### `NaN`s in the `hp` variable
There is also a significant number of observations with `NaN` / `null` value for the `hp` variable. Instead of removing these observations from the dataset, we can just set a *sentinel value* to identify these observations. Since Pokemon stats cannot be a negative number, let's represent observations with no value for these variables as -1.

We will use two different approaches in setting a sentinel value for these observations. Let's first create two copies of `pokemon_df` using the [`copy()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html) function.

In [37]:
pokemon_df1 = pokemon_df.copy()
pokemon_df2 = pokemon_df.copy()

Use the [`loc()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) function to select and assign -1 as value for the `hp` variable of these observations.

In [38]:
pokemon_df1.loc[pokemon_df1['hp'].isnull(), 'hp'] = -1

Let's check if there are any `NaN` / `null` values remaining for the `hp` variable.

In [39]:
pokemon_df1['hp'].isnull().any()

np.False_

Alternatively, we can also use the [`fillna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) function:

In [40]:
pokemon_df2['hp'] = pokemon_df2['hp'].fillna(-1)

Let's check if there are any `NaN` / `null` values remaining for the `hp` variable.

In [41]:
# Write your code here
pokemon_df1['hp'].isnull().any()

np.False_

Let's update the main copy of the Pokemon dataframe with the cleaned `hp` variable.

In [42]:
pokemon_df['hp'] = pokemon_df1['hp']

Let's count number of observations with the sentinel value.

In [43]:
# Write your code here
pokemon_df.loc[pokemon_df['type_2'] == 'None', 'type_2'].shape

(0,)

There are more than 100 instances with a sentinel value of -1 for the `hp` variable.

### `NaN`s in the `defense` variable
Write the code below to replace missing values for the `defense` variable with -1. You may use any of the approaches previously stated.

In [67]:
# Write your code here


Let's count number of observations with the sentinel value.

In [68]:
# Write your code here


There are more than 100 instances with a sentinel value of -1 for the `defense` variable.

## Further reading
**Motivation for setting a sentinel value (as opposed to leaving it as NaN)**
- This [article](https://www.residentmar.io/2016/06/12/null-and-missing-data-python.html) goes through the motivation for setting a sentinel value in much detail as well as the alternative approaches we may use in Python.